### PR TITLE
Slasher committees cache

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

# Description

Add committees cache to reduce rpc calls to archival node

### CODE REVIEW

**Comment by rauljordan** [comment]
Looks good @shayzluf , thanks!

File: slasher/service/data_update.go [hunk 1]
```diff
@@ -143,6 +140,29 @@ func (s *Service) detectSlashings(idxAtt *ethpb.IndexedAttestation) error {
 	return nil
 }
 
+func (s *Service) getCommittees(at *ethpb.Attestation) (*ethpb.BeaconCommittees, error) {
+	epoch := at.Data.Target.Epoch
+	committees, err := committeesCache.Get(s.context, epoch)
```
Comment (terencechain) [inline slasher/service/data_update.go]: Why are we using the service context? Should the input context be separated from that?

```suggestion
	committees, err := committeesCache.Get(ctx, epoch)
```

File: slasher/service/data_update.go [hunk 2]
```diff
@@ -143,6 +140,29 @@ func (s *Service) detectSlashings(idxAtt *ethpb.IndexedAttestation) error {
...
+		QueryFilter: &ethpb.ListCommitteesRequest_Epoch{
+			Epoch: epoch,
+		},
+	}
+	bCommittees, err := s.beaconClient.ListBeaconCommittees(s.context, committeeReq)
+	if err != nil {
+		log.WithError(err).Errorf("Could not list beacon committees for epoch %d", at.Data.Target.Epoch)
+		return nil, err
+	}
+	committeesCache.Put(s.context, epoch, bCommittees)
```
Comment (terencechain) [inline slasher/service/data_update.go]: same here

File: slasher/service/data_update.go (hunk 1 above)
Comment (rauljordan) [reply slasher/service/data_update.go]: +1 nice catch, we should pass in ctx as the first argument to this getCommittees function

File: slasher/cache/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,21 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
+
+go_library(
+    name = "go_default_library",
+    srcs = ["committees_cache.go"],
+    importpath = "github.com/prysmaticlabs/prysm/slasher/cache",
+    visibility = ["//visibility:public"],
```
Comment (0xKiwi) [inline slasher/cache/BUILD.bazel]: Should the visibility be public here?

**Comment by rauljordan** [comment]
PTAL @terencechain

**Pull request merged**